### PR TITLE
add FZ16 to fpcr_fz_mask for aarch64

### LINKED ISSUE #40151: Set_zero_subnormals ineffective for Float16
The [`set_zero_subnormals` function](https://github.com/JuliaLang/julia/blob/3071f34b16059ea7279c2af5694f902155d0806e/base/rounding.jl#L235) is working fine for `Float32/Float64` on x86 and ARM (aarch64) in Julia 1.6-rc3 but has no effect for `Float16`
```julia
julia> set_zero_subnormals(true)
true
julia> floatmin(Float64)/4
0.0
julia> floatmin(Float32)/4
0.0f0
julia> floatmin(Float16)/4
Float16(1.526e-5)
```
This is an issue as a64fx (tested on [Isambard](https://gw4.ac.uk/isambard/) and Fugaku) considerably slows down with subnormals:
```julia
julia> A = floatmin(Float16)*rand(Float16,1000,1000)
julia> B = floatmin(Float16)*rand(Float16,1000,1000)
julia> @btime +($A,$B);
  7.116 ms (2 allocations: 1.91 MiB)
```
compared to computations without subnormals
```julia
julia> A = rand(Float16,1000,1000)
julia> B = rand(Float16,1000,1000)
julia> @btime +($A,$B);
  274.681 μs (2 allocations: 1.91 MiB)
```

(@vchuravy I believe you've switched julia's Float16 to LLVM's half for 1.6?) @samhatfield

### PR DESCRIPTION
**Pull request by vchuravy**
Fixes #40151 by also setting the FZ16 mask

@milankl can you test on your A64fx machine?

### CODE REVIEW

**Comment by yuyichao** [comment]
Have you checked if it is defined to write to reserved bits on chips without the fp16 feature? (I'll assume you are writing the correct bits)

**Comment by milankl** [comment]
Works like a charm on a64fx

**Comment by milankl** [comment]
And last check, this PR also addresses the performance issues with subnormals and yields a decent 4x speed-up compared to Float64 without and with subnormals, whereas the latter previously slowed down things by 5x
```julia
julia> benchmark_float16_subnormals(100_000)
Without subnormals
  Float64:   88.040 μs (0 allocations: 0 bytes)
  Float32:   44.390 μs (0 allocations: 0 bytes)
  Float16:   22.351 μs (0 allocations: 0 bytes)
With 1% subnormals
  Float64:   88.240 μs (0 allocations: 0 bytes)
  Float32:   44.360 μs (0 allocations: 0 bytes)
  Float16:   22.400 μs (0 allocations: 0 bytes)
```
this just benchmarks adding two vectors{Float64/32/16} of length 100_000 element-wise.

**Comment by vchuravy** [comment]
> Have you checked if it is defined to write to reserved bits on chips without the fp16 feature? (I'll assume you are writing the correct bits)

The resource I based this off was https://developer.arm.com/documentation/100403/0201/register-descriptions/advanced-simd-and-floating-point-registers/aarch64-register-descriptions/fpcr--floating-point-control-register

If we want to be safe I can guard this behind a check for FP16 support (i.e. ID_AA64PFR0_EL1.)

**Comment by yuyichao** [comment]
> If we want to be safe I can guard this behind a check for FP16 support (i.e. ID_AA64PFR0_EL1.)

I couldn't find anything that says that the reserved bit is safe to access so I think we should safeguard this. Initializing a variable holding this flag is probably OK.
The system register isn't accessible from EL0 (as the name suggests), it's the `fullfp16` feature.

File: src/processor_arm.cpp [hunk 1]
```diff
@@ -1833,7 +1835,10 @@ extern "C" JL_DLLEXPORT int32_t jl_get_zero_subnormals(void)
 extern "C" JL_DLLEXPORT int32_t jl_set_zero_subnormals(int8_t isZero)
 {
     uint32_t fpcr = get_fpcr_aarch64();
-    fpcr = isZero ? (fpcr | fpcr_fz_mask) : (fpcr & ~fpcr_fz_mask);
+    uint32_t mask = fpcr_fz_mask;
```
Comment (yuyichao) [inline src/processor_arm.cpp:1838]: `static mask = fpcr_fz_mask | (jl_test_cpu_feature(JL_AArch64_fullfp16) ? fpcr_fz16_mask : 0);`

**Comment by vchuravy** [comment]
@milankl would be good to check whether this still works  on A64FX?

**Comment by milankl** [comment]
No problem! Will do that tomorrow

**Comment by milankl** [comment]
@vchuravy Yes, still works
```julia
julia> [floatmin(T)/2 for T in [Float64,Float32,Float16]]
3-element Vector{AbstractFloat}:
 1.1125369292536007e-308
 5.877472f-39
  Float16(3.05e-5)

julia> set_zero_subnormals(true)
true

julia> [floatmin(T)/2 for T in [Float64,Float32,Float16]]
3-element Vector{AbstractFloat}:
 0.0
 0.0f0
  Float16(0.0)
```

**Pull request merged**